pkg/spoke/addon/lease_controller.go

```diff
@@ -0,0 +1,191 @@
+	addonclient "github.com/open-cluster-management/api/client/addon/clientset/versioned"
+	addoninformerv1alpha1 "github.com/open-cluster-management/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "github.com/open-cluster-management/api/client/addon/listers/addon/v1alpha1"
+	"github.com/open-cluster-management/registration/pkg/helpers"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	coordinformers "k8s.io/client-go/informers/coordination/v1"
+	coordlisters "k8s.io/client-go/listers/coordination/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	//TODO add this to ManagedClusterAddOn api
+	addOnAvailableConditionType = "ManagedClusterAddOnConditionAvailable"
+	//TODO add this to ManagedClusterAddOn api
+	leaseDurationSeconds = 60
+	leaseDurationTimes   = 5
+)
+
+// managedClusterAddOnLeaseController udpates managed cluster addons status on the hub cluster through watching the managed
+// cluster status on the managed cluster.
+type managedClusterAddOnLeaseController struct {
+	clusterName string
+	addOnClient addonclient.Interface
+	addOnLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	leaseLister coordlisters.LeaseLister
+}
+
+// NewManagedClusterAddOnLeaseController returns an instance of managedClusterAddOnLeaseController
+func NewManagedClusterAddOnLeaseController(clusterName string,
+	addOnClient addonclient.Interface,
+	addOnInformer addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	leaseInformer coordinformers.LeaseInformer,
+	resyncInterval time.Duration,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterAddOnLeaseController{
+		clusterName: clusterName,
+		addOnClient: addOnClient,
+		addOnLister: addOnInformer.Lister(),
+		leaseLister: leaseInformer.Lister(),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(c.queueKeyFunc, leaseInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(resyncInterval).
+		ToController("ManagedClusterAddOnLeaseController", recorder)
+}
+
+func (c *managedClusterAddOnLeaseController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	if queueKey == factory.DefaultQueueKey {
+		addOns, err := c.addOnLister.ManagedClusterAddOns(c.clusterName).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		errs := []error{}
+		for _, addOn := range addOns {
+			addOnConfig, err := getAddOnConfig(addOn)
+			if err != nil {
+				// failed to get addon configuration, ignore it.
+				continue
+			}
+			if err := c.syncSingle(ctx, addOnConfig.InstallationNamespace, addOn, syncCtx.Recorder()); err != nil {
```

Comment:
Instead of this, I recommend queueing the individual items.

Reply:
> Instead of this, I recommend queueing the individual items.

queueing the individual items allows de-duping and individual retries instead of bulk retry

Reply:
updated